docs/codeql/codeql-language-guides/customizing-library-models-for-python.rst

```diff
@@ -0,0 +1,449 @@
+
+- Finally, the kind **remote** indicates that this is considered a source of remote flow.
+
+Example: Adding flow through 're.compile'
+----------------------------------------------
+
+In this example, we'll show how to add flow through calls to `re.compile`.
+`re.compile` returns a compiled regular expression for efficient evaluation, but the pattern to be compiled is stored in the `pattern` attribute of the resulting object.
+
+.. code-block:: python
+
+  import re
+
+  let y = re.compile(pattern = x); // add value flow from 'x' to 'y.pattern'
+
+Note that this flow is already recognized by the CodeQL Python analysis, but for this example, you could use the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: summaryModel
+      data:
+        - [
+            "re",
+            "Member[compile]",
+            "Argument[0,pattern:]",
+            "ReturnValue.Attribute[pattern]",
+            "value",
+          ]
+
+
+- Since we're adding flow through a function call, we add a tuple to the **summaryModel** extensible predicate.
+- The first column, **"re"**, begins the search for relevant calls at places where the **re** package is imported.
+- The second column, **Member[compile]**, is a path leading to the function calls we wish to model.
+  In this case, we select references to the **compile** function from the `re` package.
+- The third column, **Argument[0,pattern:]**, indicates the input of the flow. In this case, either the first argument to the function call or the argument named **pattern**.
+- The fourth column, **ReturnValue.Attribute[pattern]**, indicates the output of the flow. In this case, the `pattern` attribute of the return value of the function call.
+- The last column, **value**, indicates the kind of flow to add. The value **value** means the input value is unchanged as
+  it flows to the output.
+
+Example: Adding flow through 'sorted'
+-------------------------------------------------
+
+In this example, we'll show how to add flow through calls to the built-in function **sorted**:
+
+.. code-block:: python
+
+  y = sorted(x) # add taint flow from 'x' to 'y'
+
+Note that this flow is already recognized by the CodeQL Python analysis, but for this example, you could use the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: summaryModel
+      data:
+        - [
+            "builtins",
+            "Member[sorted]",
+            "Argument[0]",
+            "ReturnValue",
+            "taint",
+          ]
+
+
+- Since we're adding flow through a function call, we add a tuple to the **summaryModel** extensible predicate.
+- The first column, **"builtins"**, begins the search for relevant calls at references to the **builtin** package.
```

Comment:
... And also otherwise undefined globals with specific names. Perhaps it would be better to _not_ mention the `builtins` package explicitly, and just say it finds references to built-in names?

Reply:
Yes, I like that.

Reply:
Hm, in the end it became somewhat different...see if you like.